### PR TITLE
PLAT-906 Big Renaming of EMF Role to Permission

### PR DESCRIPTION
**Pull request by oliver-richers**
- replaced term "authorized role" with "required permission"
  - IEmfAction.getRequiredPermission()
  - IEmfPage.getRequiredPermission()
- replaced term "role membership" with "permission ownership" or "permission assignment"

```
find -type f | egrep 'sqml$|java$' | xargs grep -i 'role' | egrep -v 'ControlElement|ZeroLength'
./platform-common/src/main/java/com/softicar/platform/common/date/ISOCalendar.java: * This uses a proleptic Gregorian calendar and is thus compatible to the MySQL
./platform-db-core/src/main/java/com/softicar/platform/db/core/dbms/mysql/DbMysqlPrivilege.java:        CREATE_ROLE("CREATE ROLE"),
./platform-db-core/src/main/java/com/softicar/platform/db/core/dbms/mysql/DbMysqlPrivilege.java:        DROP_ROLE("DROP ROLE"),
./platform-ajax/src/main/java/com/softicar/platform/ajax/simple/SimpleServletRequest.java:      public boolean isUserInRole(String arg0) {
```

### CODE REVIEW

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/WorkflowTransitionActionPermission.java [hunk 1]
```diff
@@ -56,18 +56,18 @@ public boolean test(R tableRow, IBasicUser user) {
 	 * @param user
 	 *            the user that want to execute the transition (never
 	 *            <i>null</i>)
-	 * @return true if the current user is in one of the roles or if a user that
-	 *         delegated to him is in one of the roles (or has set him as
-	 *         substitute)
+	 * @return true if the current user is in one of the permissions or if a
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/WorkflowTransitionActionPermission.java:59]: That `@return` description appears nonsensical now. Also, we should avoid "him".

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java [hunk 2]
```diff
@@ -66,11 +66,11 @@ public void handleClick(AGWorkflowVersion oldWorkflowVersion) {
 			for (AGWorkflowNode oldNode: oldNodeToNewNodeMap.keySet()) {
 				final AGWorkflowNode newNode = oldNodeToNewNodeMap.get(oldNode);
 
-				// Copy Workflow Node Actions+Action Roles
+				// Copy Workflow Node Actions+Action Permissions
 				for (AGWorkflowNodeAction oldAction: oldNode.getAllActiveWorkflowNodeActions()) {
 					final AGWorkflowNodeAction newAction = oldAction.copy().setWorkflowNode(newNode).save();
-					for (AGWorkflowNodeActionRole oldActionRole: oldAction.getAllActiveWorkflowNodeActionRoles()) {
-						oldActionRole.copy().setWorkflowNodeAction(newAction).save();
+					for (AGWorkflowNodeActionPermission oldActionPermission: oldAction.getAllActiveWorkflowNodeActionPermissions()) {
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java:72]: We should use `var` here.

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java [hunk 3]
```diff
@@ -89,9 +89,9 @@ public void handleClick(AGWorkflowVersion oldWorkflowVersion) {
 				newTransition.setTargetNode(oldNodeToNewNodeMap.get(oldTransition.getTargetNode()));
 				newTransition.save();
 
-				// Copy Workflow Transition Roles
-				for (AGWorkflowTransitionRole oldRole: oldTransition.getAllActiveWorkflowTransitionRoles()) {
-					oldRole.copy().setTransition(newTransition).save();
+				// Copy Workflow Transition Permissions
+				for (AGWorkflowTransitionPermission oldPermission: oldTransition.getAllActiveWorkflowTransitionPermissions()) {
```
Comment (alexander-schmidt) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java:93]: see above

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-emf/src/main/java/com/softicar/platform/emf/permission/EmfPermissions.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package com.softicar.platform.emf.permission;
+
+import com.softicar.platform.common.core.user.IBasicUser;
+import com.softicar.platform.db.sql.field.ISqlField;
+import com.softicar.platform.emf.EmfI18n;
+
+public class EmfPermissions {
+
+	private static final IEmfPermission<?> ANYBODY = new EmfPermission<>(EmfI18n.ANYBODY, IEmfPermissionPredicate.always());
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-emf/src/main/java/com/softicar/platform/emf/permission/EmfPermissions.java:9]: using ALWAYS and NEVER

File: platform-core-module/src/main/java/com/softicar/platform/core/module/access/permission/assignment/module/instance/matrix/user/ModulePermissionOwnershipQuery.sqml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+QUERY ModulePermissionOwnershipQuery {
+
+	OPTIONAL Boolean onlyActive
+	OPTIONAL AGUser user
+	OPTIONAL AGUuid moduleUuid
+
+	SELECT ownership.*
+	SELECT instance.*
+	FROM AGModuleInstancePermissionAssignment AS ownership
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/access/permission/assignment/module/instance/matrix/user/ModulePermissionOwnershipQuery.sqml:15]: renamed to assignment

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CorePermissions.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package com.softicar.platform.core.module;
+
+import com.softicar.platform.core.module.access.permission.EmfSystemModulePermission;
+
+public interface CorePermissions {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/CorePermissions.java:5]: renamed the latter two, cannot find another name for SUPER_USER, IMO, we should get rid of it anyway

File: platform-core-module/src/main/java/com/softicar/platform/core/module/page/service/login/PageServiceLoginAuthenticator.java [hunk 7]
```diff
@@ -113,7 +113,7 @@ public AGUser authenticate() {
 
 		// check for maintenance
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/page/service/login/PageServiceLoginAuthenticator.java:114]: changed to `check for maintenance windows`

File: platform-emf/src/main/java/com/softicar/platform/emf/permission/EmfAllPermissions.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import com.softicar.platform.common.core.user.IBasicUser;
+import com.softicar.platform.emf.EmfI18n;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+/**
+ * Aggregates several {@link IEmfPermission} objects into a conjunction.
+ * <p>
+ * This {@link IEmfPermission} only applies to a user if all contained
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-emf/src/main/java/com/softicar/platform/emf/permission/EmfAllPermissions.java:15]: removed

File: platform-integration/build.gradle [hunk 9]
```diff
@@ -1,5 +1,5 @@
 dependencies {
-	implementation project(':platform-core-module')
+	implementation project(':platform-workflow-module')
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-integration/build.gradle:2]: hmm, actually, no, I was just wondering if this was intentional ;)

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/WorkflowTransitionActionPermission.java (hunk 1 above)
Comment (oliver-richers) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/transition/WorkflowTransitionActionPermission.java:59]: changed to 
```
	 * @return <i>true</i> if the given user has this permissions or if a user
	 *         that delegated the task to the user (or defined the user as
	 *         substitute) has this permission
```

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java (hunk 2 above)
Comment (oliver-richers) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java:72]: using var everywhere in that method

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java (hunk 3 above)
Comment (oliver-richers) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/version/CreateNewWorkflowVersionAction.java:93]: using var everywhere in that method

**Review by alexander-schmidt — APPROVED** [review]

**Pull request merged**